### PR TITLE
Decouple business logic from OWIN

### PR DESCRIPTION
**Pull request by joelverhagen**
1. Separate business logic from HTTP-specific things.
2. Move HTTP-specific things to NuGet.Services.BasicSearch.
4. Delete some dead code to simplify the decoupling.

Deployed to DEV with build [3.0.1174](https://nugetbuild.cloudapp.net/viewLog.html?buildId=10481&tab=buildResultsDiv&buildTypeId=Gallery_MetadataService_ContinuousIntegrati).

@maartenba @xavierdecoster @johnataylor

### CODE REVIEW

**Comment by maartenba** [comment]
(ship-it tree)

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 1]
```diff
@@ -1,59 +1,16 @@
...
-            if (!bool.TryParse(context.Request.Query["ignoreFilter"], out ignoreFilter))
-            {
-                ignoreFilter = false;
-            }
-
-            return QuerySearch(searcherManager, q, countOnly, includePrerelease, sortBy, skip, take, feed, ignoreFilter);
-        }
-
-        public static string QuerySearch(NuGetSearcherManager searcherManager, string q, bool countOnly, bool includePrerelease, string sortBy, int skip, int take, string feed, bool ignoreFilter)
+        public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherManager, string q, bool countOnly, bool includePrerelease, string sortBy, int skip, int take, string feed, bool ignoreFilter)
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Think we need to improve parameter validation and perhaps set sane defaults when no values provided? Can be trackd in a separate issue though.

e.g. if `take < 1` set it to 20, stuff like that :)

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 2]
```diff
@@ -82,13 +39,13 @@ public static string QuerySearch(NuGetSearcherManager searcherManager, string q,
             }
         }
 
-        private static string DocumentCountImpl(IndexSearcher searcher, Query query, Filter filter)
+        static void DocumentCountImpl(JsonWriter jsonWriter, IndexSearcher searcher, Query query, Filter filter)
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Why remove the explicit `private` modifier?

File: src/NuGet.Indexing/NuGet.Indexing.csproj [hunk 3]
```diff
@@ -143,9 +139,6 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\packages\NuGetGallery.Core.3.0.413-r-master\lib\net45\NuGetGallery.Core.dll</HintPath>
     </Reference>
-    <Reference Include="Owin">
-      <HintPath>..\..\packages\Owin.1.0\lib\net40\Owin.dll</HintPath>
-    </Reference>
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/NuGet.Indexing.csproj]: :rocket:

File: src/NuGet.Indexing/ResponseFormatter.cs [hunk 4]
```diff
@@ -17,28 +20,17 @@ public static class ResponseFormatter
...
-                }
-            }
+            jsonWriter.WriteStartObject();
+            WriteInfo(jsonWriter, baseAddress, searcher, topDocs);
+            WriteData(jsonWriter, searcher, topDocs, skip, take, baseAddress, includePrerelease, includeExplanation, query);
+            jsonWriter.WriteEndObject();
         }
 
-        static void WriteInfo(JsonTextWriter jsonWriter, Uri baseAddress, NuGetIndexSearcher searcher, TopDocs topDocs)
+        static void WriteInfo(JsonWriter jsonWriter, Uri baseAddress, NuGetIndexSearcher searcher, TopDocs topDocs)
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/ResponseFormatter.cs]: Prefer explicit access modifiers

File: src/NuGet.Indexing/ServiceHelpers.cs [hunk 5]
```diff
@@ -1,62 +1,13 @@
...
-        }
-
-
-        public static void WriteResponse(IOwinContext context, Exception e)
-        {
-            ServiceHelpers.TraceException(e);
-            WriteResponse(context, HttpStatusCode.InternalServerError, "{\"error\":\"Internal server error\"}").Wait();
-        }
-
         public static void TraceException(Exception e)
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/ServiceHelpers.cs]: I'll clean this one up once the logging branch gets merged in :)

File: src/NuGet.Services.BasicSearch/ServiceEndpoints.cs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+using System.Threading.Tasks;
+using Microsoft.Owin;
+using NuGet.Indexing;
+
+namespace NuGet.Services.BasicSearch
+{
+    public class ServiceEndpoints
+    {
+        public static async Task V3SearchAsync(IOwinContext context, NuGetSearcherManager searcherManager)
+        {
```
Comment (xavierdecoster) [inline src/NuGet.Services.BasicSearch/ServiceEndpoints.cs]: Think we should have the same validation on the provided parameters inside NuGet.Indexing. See my earlier comment.

**Comment by xavierdecoster** [comment]
Some nits, can probably tackled in separate issues/PRs. Good cleanup! :shipit:

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Good idea, I'll look into improving this.

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: John asked me to keep consistent style in these existing pieces of code. I do not have a strong preference but I can see the point of keeping individual files self consistent.

File: src/NuGet.Services.BasicSearch/ServiceEndpoints.cs (hunk 6 above)
Comment (joelverhagen) [inline src/NuGet.Services.BasicSearch/ServiceEndpoints.cs]: perhaps the validation should be *only* in NuGet.Indexing. Having the same validation in two places seems challenging. We will, however, need defaults set in the REST layer.

Comment (xavierdecoster) [inline src/NuGet.Services.BasicSearch/ServiceEndpoints.cs]: I think NuGet.Indexing is the one place to start with :)

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 2 above)
Comment (xavierdecoster) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: It's just a nit, but we are explicit in all other code bases. Perhaps we can be consistent in cleaning a file up when we see this? :)

**Pull request merged**